### PR TITLE
[WIP] Support for web worker breadcrumbs

### PR DESCRIPTION
**Pull request by jmshal**
Still early, and needs some more thorough testing (mainly to make sure overriding `Worker` doesn't affect existing behaviour), however this adds support for sending breadcrumbs on some key web worker events.
- [x] Creating a worker
- [x] Sending messages to a worker
- [x] Receiving messages from a worker
- [x] Terminating a worker

screenshot 2016-08-06 20 25 22

Each worker is given an auto incrementing id, as (at least in my case) when working with Workers you usually use more than one to distribute the load. So every breadcrumb includes the associated Worker's id.

Workers seem to put up a bit of a fight when trying to intercept them. As their `message` event doesn't bubble up to the `window` object. So manual overriding is necessary.

Some methods need to be bound directly to the instance, to make it clear which worker is sending/receiving messages. It would've been nicer if IE10 supported `WeakMap`, so the `Worker.prototype.(postMessage/terminate)` methods could be enhanced instead. _But I'll probably investigate a bit further into potential alternatives._

See #166

### CODE REVIEW

File: src/bugsnag.js [hunk 1]
```diff
@@ -410,6 +410,63 @@
...
   }
 
+  function trackWorkers() {
+    if(!getBreadcrumbSetting("autoBreadcrumbsWorkers", true)) {
+      return;
+    }
+
+    var workers = 0;
+
+    window.Worker = hijackConstructor(window.Worker, function(url) {
```
Comment (wordofchristian) [inline src/bugsnag.js]: I'm worried hijacking this constructor might be a little too invasive. We do it for console.log and I already don't like how that affects line numbers. We're considering creating an "examples" folder in the repo or a cookbook wiki page to demonstrate advanced customizations and breadcrumb use cases.

Comment (jacobmarshall) [inline src/bugsnag.js]: Yeah, I completely agree with you. You didn't see my face when writing this (🙃). Do you mean you're considering creating some docs which will show how to *manually* hijack (for instance - in this case) breadcrumbs into workers?

Comment (jacobmarshall) [inline src/bugsnag.js]: Idk, something like like...

```js
const worker = new Worker('./my-happy-little-worker.js');
bugsnag.attachBreadcrumbs(worker); // would setup subsequent events/method overrides
worker.postMessage('dance 💃'); // auto-breadcrumb
```

The thing is, going back to the `console.log` issue. If any of these methods (`postMessage` or `terminate`) errored, the stacktrace should still be valid, but the console would just show the top frame as coming from bugsnag.js. Which is arguably not as terrible - but I do understand it's still not favourable. We get this already with `setTimeout` and `requestAnimationFrame` anyway.

File: src/bugsnag.js [hunk 2]
```diff
@@ -410,6 +410,63 @@
...
+
+      enhance(this, "postMessage", function(message) {
+        trackWorkerBreadcrumb("Message to worker", {
+          message: message
+        });
+      });
+
+      enhance(this, "terminate", function() {
+        trackWorkerBreadcrumb("Terminated worker", {
+          uptime: (Date.now() - startTime) + "ms"
```
Comment (wordofchristian) [inline src/bugsnag.js]: Should be ` new Date().getTime()`

File: src/bugsnag.js [hunk 3]
```diff
@@ -930,6 +987,29 @@
...
+      return instance;
+    };
+    NewConstructor.prototype = Constructor.prototype;
+    NewConstructor.prototype.constructor = NewConstructor.constructor = NewConstructor;
+    return NewConstructor;
+  }
+
+  function serialiseGenericObject(obj) {
+    var str;
+    str = JSON.stringify(obj, null, "  ");
```
Comment (wordofchristian) [inline src/bugsnag.js]: This could be combined with previous line

File: src/bugsnag.js (hunk 1 above)
Comment (wordofchristian) [inline src/bugsnag.js]: I'm coming around the way you have it currently. Internally, we currently disable the console.log breadcrumb in development to get around the line number problem. I'm actually starting to wonder if the default value for `autoBreadcrumbsConsole` and `autoBreadcrumbsWorker` should be based on the release stage.

**Comment by wordofchristian** [comment]
What are your opinions on this @kattrali, @eanakashima, @snmaynard

File: src/bugsnag.js (hunk 1 above)
Comment (jacobmarshall) [inline src/bugsnag.js]: That actually makes a lot of sense, however as long as it's clear that's the case so you guys don't get a million reports of "some breadcrumbs aren't showing" 😜

That or documenting and recommending what you guys do internally as a solution? Kinda ugly, but much more open and less magic.

File: src/bugsnag.js (hunk 2 above)
Comment (jacobmarshall) [inline src/bugsnag.js]: All browsers which support workers, support `Date.now()`, but I can change it for consistency sake?

**Comment by kattrali** [comment]
I think for this to be most usable, having it off in development (along with the console ones) is probably the best option. The line number manipulation has come up a few times as troublesome in debugging, so it would probably be best to not make it worse. In general though I think this is a good feature and very useful.

**Comment by bengourley** [comment]
@jmshal we're going to close this off and ensure it either happens by default in the next major version OR is easily pluggable.

**Pull request closed without merging**